minibear/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -0,0 +1,954 @@
+
+public class DrivetrainSubsystem extends SubsystemBase {
+  /**
+   * The maximum voltage that will be delivered to the drive motors.
+   * <p>
+   * This can be reduced to cap the robot's maximum speed. Typically, this is useful during initial testing of the robot.
+   */
+  public static final double MAX_VOLTAGE = 12.0 * Constants.DriveVoltageScalar;
+
+  /**
+   * The maximum velocity of the robot in meters per second.
+   * <p>
+   * This is a measure of how fast the robot should be able to drive in a straight line.
+   * From: https://docs.wcproducts.com/wcp-swervex/general-info/ratio-options
+   * Gear ratio: 7.85:1. Free speed of 14.19 ft/s = 4.3251 m/s
+   */
+
+   //4.3251
+  public static final double MAX_VELOCITY_METERS_PER_SECOND = 2.3251;
+  public static final double MAX_ACCELERATION_METERS_PER_SECOND_SQUARED = 6.0;
+
+  public static final double MIN_VELOCITY_BOUNDARY_METERS_PER_SECOND = MAX_VELOCITY_METERS_PER_SECOND * 0.14; // 0.14 a magic number based on testing
+  /**
+   * The maximum angular velocity of the robot in radians per second.
+   * <p>
+   * This is a measure of how fast the robot can rotate in place.
+   */
+  // Here we calculate the theoretical maximum angular velocity. You can also replace this with a measured amount.
+  public static final double MAX_ANGULAR_VELOCITY_RADIANS_PER_SECOND = MAX_VELOCITY_METERS_PER_SECOND /
+          Math.hypot(DRIVETRAIN_TRACKWIDTH_METERS / 2.0, DRIVETRAIN_WHEELBASE_METERS / 2.0);
+  public static final double MIN_ANGULAR_VELOCITY_BOUNDARY_RADIANS_PER_SECOND = MAX_ANGULAR_VELOCITY_RADIANS_PER_SECOND * 0.06; // 0.06 a magic number based on testing
+  private double MAX_ANGULAR_ACCELERATION_RADIANS_PER_SECOND_SQUARED = 10.0;
+
+  private static final int PositionHistoryWindowTimeMilliseconds = 5000;
+  private static final int CommandSchedulerPeriodMilliseconds = 20;
+  private final double deltaTimeSeconds = 0.02; // 20ms scheduler time tick
+  private static final int CommandSchedulerCyclesPerSecond = 1000/CommandSchedulerPeriodMilliseconds;
+  private static final int PositionHistoryStorageSize = PositionHistoryWindowTimeMilliseconds/CommandSchedulerPeriodMilliseconds;
+
+  private static final Lock theLock = new ReentrantLock();
+
+  private final SwerveDriveKinematics swerveKinematics = new SwerveDriveKinematics(
+          // Front left
+          new Translation2d(DRIVETRAIN_TRACKWIDTH_METERS / 2.0, DRIVETRAIN_WHEELBASE_METERS / 2.0),
+          // Front right
+          new Translation2d(DRIVETRAIN_TRACKWIDTH_METERS / 2.0, -DRIVETRAIN_WHEELBASE_METERS / 2.0),
+          // Back left
+          new Translation2d(-DRIVETRAIN_TRACKWIDTH_METERS / 2.0, DRIVETRAIN_WHEELBASE_METERS / 2.0),
+          // Back right
+          new Translation2d(-DRIVETRAIN_TRACKWIDTH_METERS / 2.0, -DRIVETRAIN_WHEELBASE_METERS / 2.0)
+  );
+
+  // private byte navxSampleRate = InstalledHardware.navx2Installed? (byte) 50 : (byte) 200;
+  private byte navxSampleRate = (byte) 50;
```

Comment:
Good ... that article says 50 or 66 is right here.  According to the article, any other values just add drama.